MTSRE-388 | add support for validator assets and add candidates for 0007 validator

yashvardhan-kukreja opened the pull request:
Signed-off-by: Yashvardhan Kukreja <ykukreja@redhat.com>

It's tough to write succeeding/failing candidates for validators which involving fiddling with static assets like CSV and what not. Hence, having a `internal/testdata/assets/am{validator-id}` dedicated to such scenarios to store any number of static assets if need be for the testing of a validator.

For example: This PR itself depicts how 0007 validator's candidates make use of csv.yaml manifest provided under `internal/testdata/assets/am0007/` for its candidates.

sugarraysam commented on pkg/validators/helpers.go:
```diff
@@ -1,3 +1,35 @@
+	if err != nil {
+		return unstructured.Unstructured{}, err
+	}
+	var csvObj unstructured.Unstructured
```
Can we load other k8s objects with `unstructured.Unstructured`? The name `csvObj` suggests it would only be a csv.

yashvardhan-kukreja replied:
Ohh yes, forgot to rename this haha. Will do :)

sugarraysam commented on pkg/validators/test_am0007.go:
```diff
@@ -20,10 +22,63 @@ func (v TestAM0007) Run(mb utils.MetaBundle) (bool, string, error) {
+				InstallMode: "something",
+			},
+			Bundles: []registry.Bundle{
+				*registry.NewBundle("random-bundle", &registry.Annotations{}, &csvObj),
```
Great idea. Thinking we could have a `testutils.NewBundle(<path_to_csv>)` helper, given a path, and it can `log.Fatal` like `testutils.DefaultSucceedingCandidates` on errors. WDYT?

yashvardhan-kukreja replied:
Yep, sounds good. Done :)

sugarraysam commented on pkg/validators/helpers.go:
```diff
@@ -1,3 +1,35 @@
+
 func StringToPtr(s string) *string { return &s }
+
+func yamlToDynamicObj(yamlPath string) (unstructured.Unstructured, error) {
```
nit: Would move to `utils` package as it's not really part of validators.

yashvardhan-kukreja replied:
Gotten rid of this entire file xD

sugarraysam reviewed (approved):
Great stuff. Thank you for this! Left some nits but lgtm.

yashvardhan-kukreja commented on pkg/validators/assets/am0007/csv.yaml:
```diff
@@ -0,0 +1,92 @@
+apiVersion: operators.coreos.com/v1alpha1
```
Sure, makes sense :)

yashvardhan-kukreja commented:
@sugarraysam addressed all your suggestions. You can checkout the `testutils.NewBundle(....)` helper and if it looks good to you, then please go ahead with merging this PR. Thanks :)

sugarraysam commented on internal/testutils/testutils.go:
```diff
@@ -45,3 +75,15 @@ func DefaultSucceedingCandidates() []types.MetaBundle {
 	return res
 }
+
+func NewBundle(name string, yamlFilePaths ...string) registry.Bundle {
```
Nice !

One more thing, are we going to load bundles from assets located in a different directory? Thinking we could add `path.Join(BundleAssetsDir, "amXXXX/<filename>.yaml")` inside `NewBundle`

and maybe provide  a comment

```go
// NewBundle - create bundle from assets in `internal/testdata/assets/`
```

yashvardhan-kukreja replied:
Good stuff! Let me fix this :)

yashvardhan-kukreja replied:
> path.Join(BundleAssetsDir, path) inside NewBundle, where path is relative (e.g.: "am0007/csv.yaml")

Wouldn't this a bit confusing ( kinda over-abstracted) for the validator developer?
Because, I guess that from the eyes and perspective of the validator developer, while coding out the `test_am0007.go` they would be working inside the `pkg/validators/` directory in that specific moment. And from their eyes, the assets files would sit inside "../../internal/testdata/assets/am0007/" and that's the first thing which will pop in their mind to feed as the yamlFilePaths.
But just "am0007/*" would feel weird unless they exactly know that behind the scenes, the `NewBundle` function is prefixing their input path with "../../internal/testdata/assets". 

Sounds a lot for the validator developer (considering myself to be in the shoes of one) but wdyt :)

sugarraysam replied:
that sounds perfectly reasonable. Lets merge it ;)
